include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -53,6 +57,47 @@ __pattern_walk_n(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Function
     }
 }
 
+#if _ONEDPL_CPP20_RANGES_PRESENT
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_for_each
+//---------------------------------------------------------------------------------------------------------------------
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    auto __f_1 = 
+        [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+    oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), __f_1,
+                                                            oneapi::dpl::views::all(std::forward<_R>(__r)));
+
+    return std::ranges::borrowed_iterator_t<_R>(std::ranges::begin(__r) + std::ranges::size(__r));
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_transform
+//---------------------------------------------------------------------------------------------------------------------
+template<typename _BackendTag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F,
+         typename _Proj>
+auto
+__pattern_transform(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    auto __unary_op = [=](auto&& __val) -> decltype(auto)
+        { return std::invoke(__op, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+
+    oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), 
+            oneapi::dpl::__internal::__transform_functor<decltype(__unary_op)>{std::move(__unary_op)},
```

Comment:
Can we not write this ugly `decltype`? I hope CTAD should help us here. If __tranform_functor is an aggregate, let's write our deduction guide. What do you think?

Reply:
We have not deduction guides on oneDPL... If we want to introduce that I think it can be done later in a separate PR.

Reply:
We have the same usage (with explicit type passing) of transform_functor in many other places.
Here it is written consistency.
For now I suggest creating a ref issue to simplify the all entries

Reply:
https://github.com/oneapi-src/oneDPL/issues/1801